pkg/tfbridge/provider.go

```diff
@@ -309,12 +307,148 @@ func (p *Provider) GetSchema(ctx context.Context,
+	}
+
+	// Perform validation of the config state so we can offer nice errors.
+	warns, errs := p.tf.Validate(config)
+	for _, warn := range warns {
+		if err = p.host.Log(ctx, diag.Warning, "", fmt.Sprintf("provider config warning: %v", warn)); err != nil {
+			return nil, err
+		}
+	}
+
+	if len(errs) > 0 {
+		return nil, errors.Wrap(multierror.Append(nil, errs...), "could not validate provider configuration")
+	}
+
+	return &pulumirpc.CheckResponse{Inputs: req.GetNews()}, nil
 }
 
 // DiffConfig diffs the configuration for this Terraform provider.
 func (p *Provider) DiffConfig(ctx context.Context, req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "DiffConfig is not yet implemented")
+	urn := resource.URN(req.GetUrn())
+	label := fmt.Sprintf("%s.DiffConfig(%s)", p.label(), urn)
+	glog.V(9).Infof("%s executing", label)
+
+	// There is no logic in the TF provider that suggests that provider level config
+	// should force a new provider. Therefore, we are going to do this based on our
+	// own schema overrides. We should use ForceNew as part of the SchemaInfo to do this
+
+	var olds resource.PropertyMap
+	var err error
+	if req.GetOlds() != nil {
+		olds, err = plugin.UnmarshalProperties(req.GetOlds(), plugin.MarshalOptions{
+			Label: fmt.Sprintf("%s.olds", label), KeepUnknowns: true})
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	news, err := plugin.UnmarshalProperties(req.GetNews(), plugin.MarshalOptions{
+		Label: fmt.Sprintf("%s.news", label), KeepUnknowns: true, SkipNulls: true})
+	if err != nil {
+		return nil, err
+	}
+
+	var diffs, replaces []string
+	for _, key := range olds.StableKeys() {
+		// check that the old key exists in the new keys - if not, then report as a diff
+		// check if the new key and the old key are difference
+		if news[key].IsNull() || news[key] != olds[key] {
+			// we need to find out what the schema value of this is and report it as expected
+			if p.info.Config != nil && p.info.Config[string(key)] != nil {
+				config := p.info.Config[string(key)]
+				// now is it a ForceNew or not
+				if config.ForceNew != nil && *config.ForceNew {
+					replaces = append(replaces, string(key))
+				} else {
+					diffs = append(diffs, string(key))
+				}
+			}
+		}
+	}
+
+	for _, key := range news.StableKeys() {
+		// Check to see if the new key is not present in the old keys
+		if olds[key].IsNull() {
+			if p.info.Config != nil && p.info.Config[string(key)] != nil {
+				config := p.info.Config[string(key)]
+				// now is it a ForceNew or not
+				if config.ForceNew != nil && *config.ForceNew {
+					replaces = append(replaces, string(key))
+				} else {
+					diffs = append(diffs, string(key))
+				}
+			}
+		}
+	}
+
+	if len(diffs) > 0 || len(replaces) > 0 {
+		return &pulumirpc.DiffResponse{
+			Changes:  pulumirpc.DiffResponse_DIFF_SOME,
+			Diffs:    diffs,
+			Replaces: replaces,
```

Comment:
Can/should we return a DetaildDiff here?

@pgavlin what is the current trade off in user experience if we don’t return a DetailedDiff?

Reply:
The diff will be recomputed by the CLI and may show false diffs or omit actual diffs.